lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.cpp

```diff
@@ -1880,60 +1849,40 @@ ThreadSP ProcessGDBRemote::SetThreadStopInfo(
               StopInfo::CreateStopReasonVForkDone(*thread_sp));
           handled = true;
         }
-      } else if (!signo) {
-        addr_t pc = thread_sp->GetRegisterContext()->GetPC();
-        lldb::BreakpointSiteSP bp_site_sp =
-            thread_sp->GetProcess()->GetBreakpointSiteList().FindByAddress(pc);
-
-        // If a thread is stopped at a breakpoint site, set that as the stop
-        // reason even if it hasn't executed the breakpoint instruction yet.
-        // We will silently step over the breakpoint when we resume execution
-        // and miss the fact that this thread hit the breakpoint.
-        if (bp_site_sp && bp_site_sp->ValidForThisThread(*thread_sp)) {
-          thread_sp->SetStopInfo(StopInfo::CreateStopReasonWithBreakpointSiteID(
-              *thread_sp, bp_site_sp->GetID()));
-          handled = true;
-        }
       }
 
       if (!handled && signo && !did_exec) {
         if (signo == SIGTRAP) {
           // Currently we are going to assume SIGTRAP means we are either
           // hitting a breakpoint or hardware single stepping.
-          handled = true;
-          addr_t pc =
-              thread_sp->GetRegisterContext()->GetPC() + m_breakpoint_pc_offset;
-          lldb::BreakpointSiteSP bp_site_sp =
-              thread_sp->GetProcess()->GetBreakpointSiteList().FindByAddress(
-                  pc);
 
-          if (bp_site_sp) {
-            // If the breakpoint is for this thread, then we'll report the hit,
-            // but if it is for another thread, we can just report no reason.
-            // We don't need to worry about stepping over the breakpoint here,
-            // that will be taken care of when the thread resumes and notices
-            // that there's a breakpoint under the pc.
-            if (bp_site_sp->ValidForThisThread(*thread_sp)) {
-              if (m_breakpoint_pc_offset != 0)
-                thread_sp->GetRegisterContext()->SetPC(pc);
-              thread_sp->SetStopInfo(
-                  StopInfo::CreateStopReasonWithBreakpointSiteID(
-                      *thread_sp, bp_site_sp->GetID()));
-            } else {
-              StopInfoSP invalid_stop_info_sp;
-              thread_sp->SetStopInfo(invalid_stop_info_sp);
-            }
+          // If we were stepping then assume the stop was the result of the
```

Comment:
The previous algorithm reversed this assumption.  It first checked for the breakpoint, and if a breakpoint was at the site, then returned either a breakpoint hit (if it was for this thread) or no reason, and only if there wasn't a breakpoint there did it return the trace result.

There was also another case that I don't see handled here anymore, if we weren't at a breakpoint, and we weren't stepping, then we made a "StopReasonSignal.

I don't understand why it was right to switch the order of attributing causes here, or why we no longer need the Signal fallback.

Reply:
I was trying to detect the case where we're at a breakpoint site, but haven't executed it yet.  

The only information available at this point is that we stopped with a signal, and we know the thread's resume state, eStateStepping.  We can look at the pc and see if there is a BreakpointSite there (possibly needing to decrement the pc on Intel systems where it's reported after the breakpoint instruction has executed), and assume that we might have stopped by hitting the breakpoint.

I assume having a signal means that the thread stopped for some reason, it wasn't merely executing along at happens to be at a BreakpointSite when another thread stops.  Maybe receiving an asynchronous signal on this thread, or a debugger interrupting execution exactly when we're at a BreakpointSite but haven't executed it yet could happen.

The way I've rewritten it has the obvious problem that you are stopped at a BreakpointSite.  You instruction step, hitting the breakpoint.  This code would (now) prioritize the fact that you were stepping, and report that as the stop reason even though you really stopped by hitting the breakpoint, and the step should have stayed on the thread plan stack.

Doing "assume breakpoint, if at BreakpointSite" first means this one branch of `SetThreadStopInfo` is using the old breakpoint logic:  We report hitting the breakpoint if we're at a breakpoint site, possibly overwriting the actual stop reason.  In this case, the only stop reasons we can overwrite are:  (1) stopped by async signal, or (2) stopped by completing an instruction step.

Testing this is a bit difficult because debugserver and lldb-server will report breakpoint hits as such, and not fall through to this "we got a signal, don't know what it means" logic.

I guess in this case preserving the behavior, and possibly overwriting an asynchronous signal or an instruction stop stop-reason is the safest thing to do.

Reply:
Adding this comment at the top 
```
          // We can't disambiguate between stepping-to-a-breakpointsite and
          // hitting-a-breakpointsite. 
          //
          // A user can instruction-step, and stopped at a BreakpointSite.  
          // Or a user can be sitting at a BreakpointSite, 
          // instruction-step which hits the breakpoint and the pc does not 
          // advance.
          //
          // In both cases, we're at a BreakpointSite when stopped, and 
          // the resume state was eStateStepping.
```